benchmarks/moe.py

```diff
@@ -0,0 +1,129 @@
+import argparse
+import utils
+import logging
+import math
+import time
+
+import torch
+import torch.distributed as dist
+import torch.multiprocessing as mp
+
+from golden_configs.lm_wikitext2 import MOE as MOEConfig
+from torch.nn.parallel import DistributedDataParallel as DDP
+
+MPI_PORT = 29500
+
+def benchmark_single_process(config_class, args):
+    """Benchmark a given model using a single process and multiple devices."""
+
+    world_size = torch.cuda.device_count() if torch.cuda.is_available() else 1
+    assert world_size > 0
+    benchmark_config = utils.create_benchmark_config(args.model_name, config_class)
+    model_specs = utils.get_model_specs(args.model_name, config_class)
+
+    mp.spawn(train,
+        args=(world_size, benchmark_config, model_specs, args),
+        nprocs=world_size,
+        join=True)
+
+
+def train(rank, world_size, benchmark_config, model_specs, args):
+    logger = mp.log_to_stderr()
+    logger.setLevel(logging.DEBUG if args.debug else logging.INFO)
+    utils.init_random_seed(rank)
+
+    init_method_pgroup = "tcp://localhost:{}".format(MPI_PORT)
+    torch.distributed.init_process_group(
+            backend="gloo", rank=rank, world_size=world_size, init_method=init_method_pgroup)
```

Comment:
Lets use the nccl backend since we have standardized on it?

Reply:
done